src/createElement.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * 일반 태그이거나 함수형 컴포넌트면 createElement 함수 실행
+ */
+export const createElement = (type: string | Function, props: JSX.Props, ...children: any[]) => {
+  // 함수형 컴포넌트면 props랑 전개한 children 내려줘서 재귀적으로 createElement 실행
+  if (typeof type === 'function') {
+    return type(props, ...children);
+  }
+
+  return {
+    type,
+    props: {
+      ...props,
+      children:
+        children.length === 1 // children이 하나만 있으면 배열 말고 객체로 처리
+          ? validTextNode(children[0])
+          : children.map((child) => validTextNode(child)), // children이 여러개면 배열로 묶어서 처리
+    },
+  };
+};
+
+/**
+ * leaf 노드에 있는 텍스트 노드면 type: createTextNode로 처리
+ */
+const validTextNode = (child: any) => {
+  if (typeof child === 'string' || typeof child === 'number' || typeof child === 'boolean') {
```

Comment:
```suggestion
  if (typeof child === 'string' || typeof child === 'number') {
```
디테일한 처리 너무 좋네요👍👍👍 

원래 `boolean` 값은 리액트에서 무시되는 값으로 알고 있는데 직접 구현하실 때는 텍스트 노드로 처리해주신 이유가 있으신가요??

Reply:
아하 그렇군요! boolean 값이 무시된다는 내용을 몰라서 이렇게 했는데 수정이 필요하겠네요 :)

사실 좋은 방식이라고 생각되는 건 'child가 HTMLElement가 아닌 경우'로 타입을 좁힐 수 있으면 좋았을텐데, JSX 구문 타입 정의에 대한 이해가 부족해서 이렇게 처리한거긴 했어요ㅠㅠ

Reply:
> 아하 그렇군요! boolean 값이 무시된다는 내용을 몰라서 이렇게 했는데 수정이 필요하겠네요 :)

`null`, `undefined`, `boolean`은 무시되서 아무것도 렌더링이 되지 않아서 렌더링 하려면 문자열로 변환해서 렌더링해야 하더라구요!
저도 내부 소스 코드 분석하면서 조금 더 자세히 알게되었던 것 같아요🙃
<a href="https://velog.io/@woogur29/React%EC%97%90%EC%84%9C-%EB%AC%B4%EC%8B%9C%EB%90%98%EB%8A%94-%EA%B0%92null-undefined-boolean%EB%93%A4%EC%9D%84-%EC%96%B4%EB%96%BB%EA%B2%8C-%EC%B2%98%EB%A6%AC%ED%95%A0%EA%B9%8C">React에서 무시되는 값(null, undefined, boolean)들을 어떻게 처리될까?
</a>


> 사실 좋은 방식이라고 생각되는 건 'child가 HTMLElement가 아닌 경우'로 타입을 좁힐 수 있으면 좋았을텐데, JSX 구문 타입 정의에 대한 이해가 부족해서 이렇게 처리한거긴 했어요ㅠㅠ

맞아요 타입 잡기가 쉽지 않더라구요... 종연님이 타입 고수셔서 한번 여쭤봐도 좋을 것 같아요!

Reply:
hint) HTMLElement와 instanceof 를 적절하게 활용해보시면 좋을 것 같습니다!

Reply:
이거 시도를 해봤는데, 커스텀한 createElement 함수의 type은 string과 Function만 들어오고, 첫 글자를 대문자로 하지 않은 함수를 태그 형태로 사용했을 때에도 일반 string으로 값이 들어와서 `type instanceof HTMLElement`를 했을 때 무조건 false가 나오는 것 같네요

대신 찾아보니 `if (document.createElement(type) instanceof HTMLUnknownElement)`라는 조건으로 해당 케이스에 대해 에러 처리를 할 수 있을 것 같아요! HTML Element 표준 태그가 아닌게 태그 형태로 사용되면 HTMLUnknownElement와 프로토타입 체이닝 되는 것 같습니다